src/search_controller.ts

```diff
@@ -0,0 +1,484 @@
+import { debounce, DebouncedFunc } from './utils';
+import { StateStore } from './store';
+import type { Channel } from './channel';
+import type { StreamChat } from './client';
+import type {
+  ChannelFilters,
+  ChannelOptions,
+  ChannelSort,
+  DefaultGenerics,
+  ExtendableGenerics,
+  MessageFilters,
+  MessageResponse,
+  SearchMessageSort,
+  SearchOptions,
+  UserFilters,
+  UserOptions,
+  UserResponse,
+  UserSort,
+} from './types';
+
+export type SearchSourceType = 'channels' | 'users' | 'messages' | (string & {});
+export type QueryReturnValue<T> = { items: T[]; next?: string };
+export type DebounceOptions = {
+  debounceMs: number;
+};
+type DebouncedExecQueryFunction = DebouncedFunc<(searchString?: string) => Promise<void>>;
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export interface SearchSource<T = any> {
+  activate(sourceStateOverride?: Partial<SearchSourceState>): void;
+  deactivate(): void;
+  readonly hasMore: boolean;
+  readonly hasResults: boolean;
+  readonly isActive: boolean;
+  readonly isLoading: boolean;
+  readonly items: T[] | undefined;
+  readonly lastQueryError: Error | undefined;
+  readonly next: string | undefined;
+  readonly offset: number | undefined;
+  resetState(stateOverrides?: Partial<SearchSourceState<T>>): void;
+  search(text?: string): void;
+  searchDebounced: DebouncedExecQueryFunction;
+  readonly searchQuery: string;
+  setDebounceOptions(options: DebounceOptions): void;
+  readonly state: StateStore<SearchSourceState<T>>;
+  readonly type: SearchSourceType;
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export type SearchSourceState<T = any> = {
```

Comment:
```suggestion
export type SearchSourceState<T> = {
```

Reply:
I thought that putting the default `any` would be simpler than adding it to all the mentions in the SearchController.